# Les grandes parties qui doivent composées le Prétraitement & Feature Engineering

## Sommaire

1. Import des bibliothèques
2. Chargement des données
3. Vérification initiale du dataset
4. Traitement des valeurs manquantes
5. Suppression des variables inutiles
6. Feature Engineering
7. Encodage des variables catégorielles
8. Séparation des variables explicatives et de la cible
9. Validation et sauvegarde des jeux de données 



## 1. Import des bibliothéques

Cette première étape consiste à importer les bibliothèques nécessaires au traitement, à la transformation et à la sauvegarde des données.

In [396]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import StandardScaler

## 2. Chargement des données

In [397]:
df = pd.read_csv("../data/raw/Titanic-Dataset.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Vérification 

In [398]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [399]:
df.shape

(891, 12)

Conclusion

Le dataset brut est correctement chargé. Les variables correspondent à celles étudiées lors de l'EDA. Les traitements réalisés dans ce notebook vont permettre de préparer les données pour la phase de modélisation.

## 3. Vérification initiale du dataset

Avant transformation : 

In [400]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [401]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Conclusion :

Les valeurs manquantes identifiées précédemment sont retrouvées principalement dans Age, Cabin et Embarked. Les traitements appliqués dans les prochaines étapes vont permettre de rendre le dataset compatible avec les algorithmes de Machine Learning.

## 4. Traitement des valeurs manquantes

### 1. Variable Age

Lors de l'EDA :

177 valeurs manquantes ; variable importante. --> On conserve donc Age.

Méthode recommandée :

Remplacement par la médiane.

--> Pourquoi ?
robuste aux valeurs extrêmes ; adaptée à une distribution non parfaitement normale.

In [402]:
df["Age"].fillna(df["Age"].median(), inplace=True)
df.isnull().sum()

/var/folders/c5/03kdk4yj6jdg24y7stt42psm0000gn/T/ipykernel_9583/1406721651.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Age"].fillna(df["Age"].median(), inplace=True)


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 2. Variable Embarked

Seulement quelques valeurs manquantes.
On utilise la modalité majoritaire : la valeur la plus fréquente 



In [403]:
df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)
df.isnull().sum()

/var/folders/c5/03kdk4yj6jdg24y7stt42psm0000gn/T/ipykernel_9583/1405547764.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

### 3. Variable Cabin

Seulement quelques valeurs manquantes.
On utilise la modalité majoritaire : la valeur la plus fréquente 

Lors de l'EDA :

beaucoup de valeurs manquantes.

Deux possibilités :

Option 1 : suppression
Simple mais perte d'information.

Option 2 : création d'une variable - On va conserver une information :
Le passager possède-t-il une cabine connue ?

On choisit l'option 2.

In [404]:
# Création de la colonne
df["Cabin_known"] = df["Cabin"].notna().astype(int)
# suppression de l'ancienne colonne 
df.drop("Cabin", axis=1, inplace=True)
# Vérification
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
Cabin_known    0
dtype: int64

#### Conclusion

Les valeurs manquantes ont été traitées. La variable Age a été imputée par sa médiane, Embarked par sa modalité la plus fréquente et Cabin a été transformée en une variable binaire indiquant la présence d'une information de cabine.

## 5. Suppression des variables inutiles

### PassengerId

Pourquoi ? C'est un identifiant technique.
Il n'a aucune relation logique avec la survie.


In [405]:
df.drop("PassengerId", axis=1, inplace=True)
# Vérification 
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked,Cabin_known
0,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S,0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C,1
2,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S,0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S,1
4,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S,0


### Ticket

Pourquoi ? 
-   beaucoup de modalités différentes ;
-   difficilement exploitable directement.


In [406]:
df.drop("Ticket", axis=1, inplace=True)
# Vérification
df.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Fare,Embarked,Cabin_known
0,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,7.2500,S,0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,71.2833,C,1
2,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,7.9250,S,0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,53.1000,S,1
4,0,3,"Allen, Mr. William Henry",male,35.0,0,0,8.0500,S,0


### Name 

Avant suppression, on peut extraire une information utile.
Voir Feature Engineering.

## 6. Feature Engineering 

Cette partie consiste à créer de nouvelles variables plus informatives.

### 0. Petit point méthode : 

La question n'est donc pas : **"Quelles variables puis-je créer ?"**

mais plutôt :**"Quelle information utile n'est pas encore représentée par mes variables actuelles ?"**

1. Le point de départ c'est l'EDA : 
Tu ne fais jamais du Feature Engineering avant d'avoir exploré les données. L'EDA te révèle par exemple :
-   des variables très corrélées ;
-   des distributions particulières ;
-   des interactions entre variables ;
-   des informations cachées dans une colonne.

C'est à partir de ces observations que tu proposes de nouvelles variables.

2. Les grandes familles de Feature Engineering
    a. Extraire une information cachée : Statut social par exemple
    b. Combiner plusieurs variables : Création de FamilySize
    c. Simplifier une variable : Regroupement sous Rare des statuts rares
    d. Transformer une variable numérique : par exemple au lieu de Salaire, on peut créer la variable Salaire élevé
    e. Créer des interactions : Parfois, une variable seule explique peu, mais deux variables ensemble sont très informatives : 
        Avec Sex et Pclass on peut créer par exemple Sex*Pclass

3. Quand créer une nouvelle variable 

    Pose toi ces questions : 

    a. Est-ce que cette nouvelle variable apporte une information nouvelle ?
    b. Est-ce qu'elle simplifie le problème ?
    c. Est-ce qu'elle traduit une connaissance métier ?

4. Quand ne pas faire de Feature Engineering 

    a. Si la variable est redondante : créer Prix TTC = HT + TVA si ces deux dernières sont connues ne sert à rien car cette infos est déjà disponible
    b. Si la variable est arbitraire : cela n'a pas de sens 
    c. Si cela introduit une fuite de données : Le client va-t-il quitter l'entreprise? et on souhaite créer : **Nombre de jours après la résiliation** ; pas de sens car on ne connait pas cela au moment de la prédiction. 
    d. Si cela va compliquer le modèle sans bénéfice : on ne crée pas 20 variables si 2 suffisent

5. Comment savoir si une variable est utile? 

Après l'avoir créée, tu peux :
-   refaire une petite analyse (distribution, relation avec la cible) ;
-   mesurer son importance avec un modèle (Random Forest, XGBoost) ;
-   comparer les performances du modèle avec et sans cette variable.

Une variable qui n'améliore rien peut être supprimée.



### 1. Extraction du titre de la personne

La variable Name contient une information sociale.

Exemple : Dans "Braund, Mr. Owen Harris", on extrait Mr


In [407]:
#df["Title"] = df["Name"].str.extract(", ([A-Za-z]+)\.", expand=False)

In [408]:
df["Title"] = (
    df["Name"]
    .str.extract(r",\s*([^.]*)\.")[0]
    .str.strip()
)

print(
    df["Title"].value_counts()
)

df["Title"] = df["Title"].replace({
    "Mlle": "Miss",
    "Ms": "Miss",
    "Mme": "Mrs"
})

common_titles = [
    "Miss",
    "Mr",
    "Mrs",
    "Master"
]

df["Title"] = df["Title"].where(
    df["Title"].isin(common_titles),
    "Rare"
)

print(
    df["Title"].value_counts()
)

Title
Mr              517
Miss            182
Mrs             125
Master           40
Dr                7
Rev               6
Mlle              2
Major             2
Col               2
the Countess      1
Capt              1
Ms                1
Sir               1
Lady              1
Mme               1
Don               1
Jonkheer          1
Name: count, dtype: int64
Title
Mr        517
Miss      185
Mrs       126
Master     40
Rare       23
Name: count, dtype: int64


On visualise les différentes valeurs obtenues : 

In [409]:
df["Title"].value_counts()

Title
Mr        517
Miss      185
Mrs       126
Master     40
Rare       23
Name: count, dtype: int64

Remarque : On a beaucoup trop de titre pour que cela soit exploitable, donc on regroupe les titres rares, sous une même case : 

In [410]:
rare_titles = [
    "Lady","Countess", "the Countess","Capt",
    "Col","Don","Dr","Major",
    "Rev","Sir","Jonkheer"
]

df["Title"] = df["Title"].replace(
    rare_titles,
    "Rare"
)

On visualise à nouveau les différentes valeurs obtenues : 

In [411]:
df["Title"].value_counts()

Title
Mr        517
Miss      185
Mrs       126
Master     40
Rare       23
Name: count, dtype: int64

Avec le tableau suivant de correspondances : 

| Titre      | Signification       | Utilisation                            |
| ---------- | ------------------- | -------------------------------------- |
| **Mr**     | Monsieur            | Homme adulte                           |
| **Mrs**    | Madame              | Femme mariée                           |
| **Miss**   | Mademoiselle        | Femme non mariée (ou jeune fille)      |
| **Master** | Maître              | Jeune garçon                           |
| **Ms**     | Madame/Mademoiselle | Femme sans préciser son statut marital |


On peut remplacer Mlle par Miss, Ms avec Miss par choix (on aurait pu le remplacer par Mrs), Mme par Mrs : 

In [412]:
df["Title"] = df["Title"].replace({
    "Mlle": "Miss",
    "Ms": "Miss",
    "Mme": "Mrs"
})

In [413]:
# Vérification 
df["Title"].value_counts()

Title
Mr        517
Miss      185
Mrs       126
Master     40
Rare       23
Name: count, dtype: int64

Ensuite on supprime la donnée Name : 

In [414]:
print(df.loc[759, ["Name", "Title"]])
print(df.loc[759, "Title"])
print(repr(df.loc[759, "Title"]))
df.drop("Name", axis=1, inplace=True)

Name     Rothes, the Countess. of (Lucy Noel Martha Dye...
Title                                                 Rare
Name: 759, dtype: object
Rare
'Rare'


### 2. Taille de la famille 


In [415]:
df["FamilySize"] = df["SibSp"] + df["Parch"] + 1

### 3. Voyage seul




In [416]:
df["IsAlone"] = np.where(
    df["FamilySize"] == 1,
    1,
    0
)

Pour FamilySize et IsAlone on ne supprime pas SibSp et Parch car les informations données sont complémentaires.

In [417]:
# Vérification 
df.head()

,Survived,Pclass,Sex,Age,SibSp,Parch,Fare,Embarked,Cabin_known,Title,FamilySize,IsAlone
0,0,3,male,22.0,1,0,7.2500,S,0,Mr,2,0
1,1,1,female,38.0,1,0,71.2833,C,1,Mrs,2,0
2,1,3,female,26.0,0,0,7.9250,S,0,Miss,1,1
3,1,1,female,35.0,1,0,53.1000,S,1,Mrs,2,0
4,0,3,male,35.0,0,0,8.0500,S,0,Mr,1,1


### Conclusion Feature Engineering

De nouvelles variables ont été créées afin d'améliorer l'information disponible pour les modèles. Le titre extrait du nom apporte une information sociale, tandis que la taille de famille et le fait de voyager seul permettent de mieux représenter la situation familiale des passagers.

## 7. Encodage des variables catégorielles

Les modèles ne comprennent pas les textes.
Variables concernées :
-   Sex
-   Embarked
-   Title

### 1. Variable Sex

In [418]:
df["Sex"] = df["Sex"].map({
    "male":0,
    "female":1
})

### 2. Variables à plusieurs catégories : Embarked, Title

On utilise One-Hot Encoding, mais le provlème à étudier est dans le paramètre drop_first. 

Expliquons ce que fait drop_first : 

Lorsqu'on applique un One-Hot Encoding, chaque catégorie est transformée en une colonne binaire.

Exemple avec **Embarked**:

| Embarked |
| -------- |
| C        |
| Q        |
| S        |

-   Sans drop_first : 

| Embarked_C | Embarked_Q | Embarked_S |
| ---------- | ---------- | ---------- |
| 1          | 0          | 0          |
| 0          | 1          | 0          |
| 0          | 0          | 1          |


-   Avec drop_first : 

| Embarked_Q | Embarked_S |
| ---------- | ---------- |
| 0          | 0          |
| 1          | 0          | 
| 0          | 1          | 

La catégorie C devient la catégorie de référence.

comme : Embarked_C + Embarked_Q + Embarked_S = 1 ; Les trois colonnes sont parfaitement dépendantes. Il s'agit de la multicolinéarité.

Quand utiliser le paramètre ? 

-   Pour des modèle qui estime des coefficients et que l'on souhaite éviter la multicolinéarité : Reg.Log., Reg.Lin., GLM. Dans ces modèles conserver toutes les colonnes peut rendre instable les coefficients ou difficile à interpréter.

Quand ne pas utiliser le paramètre ? 

-   Pour les modèles basés sur des arbres ou sur des règles de séparation :
    -   Decision Tree
    -   Random Forest
    -   XGBoost
    -   LightGBM
    -   CatBoost

Ces modèles ne sont pas affectés par la multicolinéarité des variables issues du One-Hot Encoding.

On va donc créer deux datasets issu du premier : un avec drop_first = false pour les modèle ne se souciant pas de la multicolinéarité; et True pour ceux qui peuvent en souffrir.








In [419]:
# Création du 1er dataset où drop_first = False
df_full_columns = pd.get_dummies(
    df, columns=["Embarked", "Title"],
    drop_first=False, dtype=int)                            

# Création du 2ème dataset où drop_first = True
df_less_columns = pd.get_dummies(
    df, columns=["Embarked", "Title"],
    drop_first=True, dtype=int)    

## 8. Séparation des variables explicatives et de la cible

**Objectif**

Les algorithmes de Machine Learning doivent distinguer :
les variables d'entrée (les caractéristiques des passagers) ;
la variable à prédire.
Dans notre étude, la variable cible est : **Survived**

Elle prend les valeurs :
0 : le passager n'a pas survécu ;
1 : le passager a survécu.
Toutes les autres colonnes constituent les variables explicatives.

### 1. Séparation pour le dataset complet(df_full_columns)

Le premier jeu de données, obtenu après un encodage complet des variables catégorielles (drop_first=False), est séparé en variables explicatives (X) et variable cible (y).



In [420]:
X_full = df_full_columns.drop(columns="Survived")
y_full = df_full_columns["Survived"]

### 2. Séparation pour le dataset diminué (df_less_columns)

Le second jeu de données, obtenu après suppression d'une catégorie de référence (drop_first=True), est également séparé en variables explicatives et variable cible.




In [421]:
X_less = df_less_columns.drop(columns="Survived")
y_less = df_less_columns["Survived"]

### 3. Vérification


In [422]:
print("Dimensions de X_less :", X_less.shape)
print("Dimensions de y_less :", y_less.shape)
print("Dimensions de X_full :", X_full.shape)
print("Dimensions de y_full :", y_full.shape)

Dimensions de X_less : (891, 15)
Dimensions de y_less : (891,)
Dimensions de X_full : (891, 17)
Dimensions de y_full : (891,)


Les deux jeux de données ont été séparés en variables explicatives (X) et variable cible (y). Cette étape est indispensable pour préparer les données à l'entraînement des modèles de classification. Les variables explicatives regroupent l'ensemble des caractéristiques des passagers, tandis que la variable cible correspond à la survie (Survived), qui sera prédite par les différents algorithmes de Machine Learning.

## 9. Validation et sauvegarde des jeux de données 


**Objectif**

Cette étape consiste à effectuer une dernière vérification des jeux de données avant leur utilisation dans les modèles de Machine Learning. L'objectif est de s'assurer que les données sont complètes, cohérentes et compatibles avec les algorithmes qui seront entraînés dans le notebook suivant.



### 1. Vérifications des dimensions

In [423]:
print("X_full :", X_full.shape)
print("y_full :", y_full.shape)
print("X_less :", X_less.shape)
print("y_less :", y_less.shape)

X_full : (891, 17)
y_full : (891,)
X_less : (891, 15)
y_less : (891,)


Conclusion

Les dimensions des variables explicatives et de la variable cible sont cohérentes. Le nombre d'observations est identique entre X et y pour chacun des deux jeux de données.

### 2. Vérifications des types de données et valeurs manquantes

In [424]:
# Vérification des types de données
X_full.dtypes

Pclass            int64
Sex               int64
Age             float64
SibSp             int64
Parch             int64
Fare            float64
Cabin_known       int64
FamilySize        int64
IsAlone           int64
Embarked_C        int64
Embarked_Q        int64
Embarked_S        int64
Title_Master      int64
Title_Miss        int64
Title_Mr          int64
Title_Mrs         int64
Title_Rare        int64
dtype: object

In [425]:
# Vérification des types de données
X_less.dtypes


Pclass           int64
Sex              int64
Age            float64
SibSp            int64
Parch            int64
Fare           float64
Cabin_known      int64
FamilySize       int64
IsAlone          int64
Embarked_Q       int64
Embarked_S       int64
Title_Miss       int64
Title_Mr         int64
Title_Mrs        int64
Title_Rare       int64
dtype: object

In [426]:
# Vérification des valeurs manquantes
X_full.isnull().sum()


Pclass          0
Sex             0
Age             0
SibSp           0
Parch           0
Fare            0
Cabin_known     0
FamilySize      0
IsAlone         0
Embarked_C      0
Embarked_Q      0
Embarked_S      0
Title_Master    0
Title_Miss      0
Title_Mr        0
Title_Mrs       0
Title_Rare      0
dtype: int64

In [427]:
# Vérification des valeurs manquantes
X_less.isnull().sum()

Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Cabin_known    0
FamilySize     0
IsAlone        0
Embarked_Q     0
Embarked_S     0
Title_Miss     0
Title_Mr       0
Title_Mrs      0
Title_Rare     0
dtype: int64

**Conclusion**

Toutes les variables explicatives sont désormais numériques (int64, float64 ou bool selon les choix d'encodage), ce qui les rend compatibles avec les algorithmes de Machine Learning.

Aucune valeur manquante n'est présente dans les jeux de données après les opérations de prétraitement.

### 3. Vérifications des doublons

In [428]:
print(X_full.duplicated().sum())
print(X_less.duplicated().sum())

119
119


Sachant que les doublons sont apparues une fois les colonnes inutiles supprimées et les encodage réalisés, il n' y a donc pas de problèmes car au départ du projet il n'y avait pas de doublons. 

### 4. Aperçu des jeux de données



In [429]:
X_full.head()

,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_known,FamilySize,IsAlone,Embarked_C,Embarked_Q,Embarked_S,Title_Master,Title_Miss,Title_Mr,Title_Mrs,Title_Rare
0,3,0,22.0,1,0,7.2500,0,2,0,0,0,1,0,0,1,0,0
1,1,1,38.0,1,0,71.2833,1,2,0,1,0,0,0,0,0,1,0
2,3,1,26.0,0,0,7.9250,0,1,1,0,0,1,0,1,0,0,0
3,1,1,35.0,1,0,53.1000,1,2,0,0,0,1,0,0,0,1,0
4,3,0,35.0,0,0,8.0500,0,1,1,0,0,1,0,0,1,0,0


In [430]:
X_less.head()

,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin_known,FamilySize,IsAlone,Embarked_Q,Embarked_S,Title_Miss,Title_Mr,Title_Mrs,Title_Rare
0,3,0,22.0,1,0,7.2500,0,2,0,0,1,0,1,0,0
1,1,1,38.0,1,0,71.2833,1,2,0,0,0,0,0,1,0
2,3,1,26.0,0,0,7.9250,0,1,1,0,1,1,0,0,0
3,1,1,35.0,1,0,53.1000,1,2,0,0,1,0,0,1,0
4,3,0,35.0,0,0,8.0500,0,1,1,0,1,0,1,0,0


### 5. Sauvegarde des datasets

In [431]:
df_full_columns.to_csv(
    "../data/processed/titanic_full_columns.csv",
    index=False
)

df_less_columns.to_csv(
    "../data/processed/titanic_less_columns.csv",
    index=False
)



### 6. Conclusion de la partie 9


Les opérations de prétraitement et de Feature Engineering ont permis d'obtenir deux jeux de données prêts pour la phase de modélisation. 

Les valeurs manquantes ont été traitées, les variables catégorielles ont été encodées et les nouvelles variables créées ont été intégrées au dataset. 

Les dernières vérifications confirment l'absence de valeurs manquantes et la cohérence des types de données. Les deux versions du jeu de données ont été sauvegardées afin d'être utilisées selon les besoins des différents modèles de Machine Learning.